File: binary_tree/python/MaximumPathSumBT.py
max_sum = 0

class Node:
    def __init__(self, val):
        self.val = val
        self.left = self.right = None

def max_path_sum(root):
    global max_sum
    if not root:    return 0
    left_sum = max_path_sum(root.left)
    right_sum = max_path_sum(root.right)
    max_single_sum = max(root.val, root.val + left_sum, root.val + right_sum)
    max_top = max(max_single_sum, root.val + left_sum + right_sum)
    max_sum = max(max_sum, max_top)
    return max_single_sum

if __name__=='__main__':
    root = Node(10) 
    root.left = Node(2)
    root.right = Node(15)
    root.left.left = Node(-4)
    root.left.right = Node(-6) 
    root.left.left.left = Node(28)
    root.left.left.right = Node(-22)
    root.right.right = Node(-25)
    root.right.right.left = Node(3)
    root.right.right.right = Node(4)
    max_path_sum(root)
    print(max_sum)

File: binary_tree/Java/MaximumpathsumBT.java

public class MaximumpathsumBT {
    public static int maxPathSum(Node root) {
        int maxValue[] = new int[1];
        maxValue[0] = Integer.MIN_VALUE;
        maxPathDown(root, maxValue);
        return maxValue[0];
    }

    public static int maxPathDown(Node node, int maxValue[]) {
        if (node == null) return 0;
        int left = Math.max(0, maxPathDown(node.left, maxValue));
        int right = Math.max(0, maxPathDown(node.right, maxValue));
        maxValue[0] = Math.max(maxValue[0], left + right + node.data);
        return Math.max(left, right) + node.data;
    }

    public static void main(String[] args) {
        Node root = new Node(10);
        root.left = new Node(2);
        root.right = new Node(15);
        root.left.left = new Node(-4);
        root.left.right = new Node(-6);
        root.left.left.left = new Node(28);
        root.left.left.right = new Node(-22);
        root.right.right = new Node(-25);
        root.right.right.left = new Node(3);
        root.right.right.right = new Node(4);
        int res = maxPathSum(root);
        System.out.println("Path Sum: " + res);
    }
}
